### PR TITLE
CI: Run tests in Cygwin for Cygwin CI

### PR DESCRIPTION
**Pull request by DWesl**
The current version installs the Cygwin base packages (but not any python versions), then runs the rest of the steps using the previously-installed Windows-native python.  This should run the tests in a Cygwin python.

### CODE REVIEW

**Review by FFY00 — APPROVED** [review]
Hi @DWesl, thank you so much for looking into this! It is highly appreciated.

The PR looks great, I just want to wait for a new GitPython release before merging so that we can be sure that is why the tests are currently failing.

**Comment by FFY00** [comment]
I think it's gonna be a while, so I am gonna rebase and merge this PR.

File: tests/test_wheel.py
Comment (rgommers) [inline tests/test_wheel.py:60]: This should be `.pyd` as well I guess? Either that or `.so` or `.dll`. But `dll.a` is the import library, not the extension module.

**Comment by DWesl** [comment]
> I think it's gonna be a while, so I am gonna rebase and merge this PR.

Should we install GitPython from the repository rather than from PyPI so the fix is included?  I think the tests still fail with the released version.

**Comment by FFY00** [comment]
We are now installing from the repo but there are still errors, I don't know what's going on. @DWesl can you try running the tests with nox on a Cygwin environment?

**Comment by dnicolodi** [comment]
@DWesl The current way to generate wheel tags from extension filenames is doomed. I think the sane approach is the one in #202. It should already work with Cygwin but IIUC our CI setup is incomplete. Do you mind if I cherry pick the CI extension commits from this PR into that one?

**Comment by DWesl** [comment]
> We are now installing from the repo but there are still errors, I don't know what's going on. @DWesl can you try running the tests with nox on a Cygwin environment?

The most recent commit passes locally, except for what I suspect are local network errors.



> @DWesl The current way to generate wheel tags from extension filenames is doomed. I think the sane approach is the one in #202. It should already work with Cygwin but IIUC our CI setup is incomplete. Do you mind if I cherry pick the CI extension commits from this PR into that one?

You are welcome to try it.

**Comment by dnicolodi** [comment]
CI is not very happy. There seems to be something wrong with `git` in Cygwin.

File: .github/workflows/tests-cygwin.yml
Comment (dnicolodi) [inline .github/workflows/tests-cygwin.yml]: The code that fails is this https://github.com/mesonbuild/meson-python/blob/ef67e0e03c3224a8703809ce6eafb8ac5ce352d5/tests/conftest.py#L33-L51 which is called for example here https://github.com/mesonbuild/meson-python/blob/ef67e0e03c3224a8703809ce6eafb8ac5ce352d5/tests/conftest.py#L96
It should not depend on CWD of the tests runner. Why it fails although is not very clear.

Comment (DWesl) [reply .github/workflows/tests-cygwin.yml]: I was going to say "GitPython hasn't done a release since they merged the changes to get it working on Cygwin", but apparently that's only true on GitHub; PyPI had two releases a month ago.

`actions/checkout` is done with Windows git, not Cygwin, so there would be (different) errors operating on the `meson-python` repository, but the errors are apparently from a brand-new repository, so that's not it either.

My next guess would be `fork()` failure since CI runs into that a decent bit, but this project doesn't have many dependencies, nor does it have many C extension modules, and also I think I've found the actual `stdout` and `stderr` in the logs, which show a different message.

There's no reasons coming to mind for why `rm -rf .git; git init; git config gpgsign false` would fail, especially with that error

Comment (dnicolodi) [reply .github/workflows/tests-cygwin.yml]: Indeed. By the way @DWesl , do you know if there is a n easy way to cache the cygwin installation done by the `cygwin/cygwin-install-action`? Debugging this on the CI having to wait for Cygwin to install at each run is a bit tedious. On a branch I setup caching for the Python dependencies, cutting down the `pip install` time considerably, but installing Cygwin still takes anything between one minute and a half and three minutes.

Comment (dnicolodi) [reply .github/workflows/tests-cygwin.yml]: The fix for the `git` issue is this: https://github.com/mesonbuild/meson-python/blob/68e1d391e5a81d588da35ebaabe908473dbcbfee/.github/workflows/tests.yml#L91-L98

File: .github/workflows/tests-cygwin.yml
Comment (dnicolodi) [inline .github/workflows/tests-cygwin.yml:37]: This is not enough. With this the installation path is populated but the installation will be run anyway. I don't think this saves any time. Anyhow, I can live with the CI job taking a couple of minutes.

File: .github/workflows/tests-cygwin.yml
Comment (DWesl) [reply .github/workflows/tests-cygwin.yml]: So it's the second issue, but git for some reason decided to have two error messages for that.  Lovely

File: .github/workflows/tests-cygwin.yml
Comment (DWesl) [reply .github/workflows/tests-cygwin.yml:37]: The installation will be run, but that cache includes the list of installed packages; the install step will only download and extract packages if there are new versions (which means I should expire the cache at least once a month).  I would very much prefer to run the installer to be sure the dlls get rebased properly (which means I need to run `rebase-trigger full` so it knows to do that).

You're probably going to have more runins with the length of the CI job; most of the ones I've built take half an hour.  Should I take this back out again?

Comment (dnicolodi) [reply .github/workflows/tests-cygwin.yml:39]: I don't think we will be merging this PR either way. The work in #202 supersedes all of it. I just have one funny test error on Cygwin to track down: https://github.com/mesonbuild/meson-python/actions/runs/3456021871/jobs/5768534392 We setup two virtual environments as part of the tests, for some reason the second time we do it, it fails. The annoying thing is that it fails with a backtrace that does not match my copy of the CPython 3.9.15 source code and I haven't been able to locate the Cygwin Python sources

Comment (DWesl) [reply .github/workflows/tests-cygwin.yml:39]: If you've got a Cygwin install, run `setup-x86_64.exe` again and check the `Src?` tick box; you may also want to install the `cygport` package to simplify the build process (`cygport python39.cygport download prep compile test install package` if you want to do a local install, using [the Cygwin source package mentioned above](https://cygwin.com/cgi-bin2/package-cat.cgi?file=x86_64%2Fpython39-src%2Fpython39-3.9.10-1-src&grep=python39-src) and [a local download repository](https://cygwin.com/package-server.html))

Comment (dnicolodi) [reply .github/workflows/tests-cygwin.yml:39]: I don't have any Windows machine were to install Cygwin. I was hoping that a git repository with the sources exists somewhere. With your indications above I found it, https://cygwin.com/git-cygwin-packages/?p=git/cygwin-packages/python39.git but it seems the sources are maintained as a set of patches... This is so '90ies... 😄

Comment (dnicolodi) [reply .github/workflows/tests-cygwin.yml:39]: Wait, the git repo above has the history only up to Python 3.9.9 but version 3.9.10 is the current one. I guess the git repo is not where the development happens.

Comment (DWesl) [reply .github/workflows/tests-cygwin.yml:39]: > I don't have any Windows machine were to install Cygwin. I was hoping that a git repository with the sources exists somewhere. With your indications above I found it, https://cygwin.com/git-cygwin-packages/?p=git/cygwin-packages/python39.git but it seems the sources are maintained as a set of patches... This is so '90ies... 😄

Well, the source tarball from python.org and a bunch of packages that haven't been accepted upstream.  I think the standardized script to apply the packages, build and install the software, and create a distribution might be the main update to the package format since it was introduced back then.  

> Wait, the git repo above has the history only up to Python 3.9.9 but version 3.9.10 is the current one. I guess the git repo is not where the development happens.

Not really; the maintainer usually does the build locally from the upstream sources then uploads the resulting packages; synchronizing the build script with the repository is new within the past few years and not everyone has updated their muscle memory.

Poking around the mirror `cygwin/cygwin-install-action` uses by default, I found https://mirrors.kernel.org/sourceware/cygwin/x86_64/release/python39/python39-3.9.10-1-src.tar.xz, which should have everything you need.

As for the venv failure; I used nox here, which probably uses some virtual environment setup for dependency management and seems to get through the "install dependencies" part just fine; running `python -m venv test-venv; . test-venv/bin/activate; python -Im ensurepip --upgrade --default-pip` worked fine as well, so I'm not sure what's going on there.

Comment (dnicolodi) [reply .github/workflows/tests-cygwin.yml:39]: Thanks. I'm looking into this. The venv creation that fails is the second one we do in code in the testsuite. See https://github.com/mesonbuild/meson-python/blob/ef67e0e03c3224a8703809ce6eafb8ac5ce352d5/tests/conftest.py#L62-L82

Comment (DWesl) [reply .github/workflows/tests-cygwin.yml:39]: That bit of the traceback looks like it was added in https://cygwin.com/git-cygwin-packages/?p=git/cygwin-packages/python39.git;a=blob;f=00189-use-rpm-wheels.patch;h=e5cf8f7831a34d4d8a0a6b543358d495cd182ace;hb=42494e325a050ba03638568d7318f8f0075e25fb
which seems to have been borrowed from Fedora or a related distribution.

Comment (DWesl) [reply .github/workflows/tests-cygwin.yml:39]: It looks like it's failing because it's not finding wheels in `/usr/share/python-wheels`; would running `python -m pip wheel --wheel-dir /usr/share/python-wheels wheel setuptools pip` help with that?

Comment (dnicolodi) [reply .github/workflows/tests-cygwin.yml:39]: Thank you for the pointers, this helps a lot! Putting these wheels there should be responsibility of the system vendor, not something we need to patch. I need to debug this a bit more. At least now I know what to look for.

Comment (dnicolodi) [reply .github/workflows/tests-cygwin.yml:39]: Lovely, `/usr/share/python-wheels/` does not even exist, despite I explicitly install the `python39-pip` and `python39-setuptools` packages

Comment (dnicolodi) [reply .github/workflows/tests-cygwin.yml:39]: @DWesl can you please check the content of `/usr/share/python-wheels/` on your Cygwin system? Also, what does it happen if you do something like `python -m venv $(mktemp -d)`?

Comment (dnicolodi) [reply .github/workflows/tests-cygwin.yml:39]: Accordingly to https://cygwin.com/cgi-bin2/package-grep.cgi there is no Cygwin package that installs something in` /usr/share/python-wheels/ ` so I am not sure I understand how the patch you linked is even supposed to work.

Comment (DWesl) [reply .github/workflows/tests-cygwin.yml:39]: > @DWesl can you please check the content of `/usr/share/python-wheels/` on your Cygwin system? 

I have a few hundred wheels in there, because I often create my own packages and make sure my packages install wheels there so I don't have to sit waiting for scipy to build every time I want to test something with tox.  I think it worked fine before, but I haven't checked.

> Also, what does it happen if you do something like `python -m venv $(mktemp -d)`?

Works the same as the other one, once I found the temporary environment.

It looks like the function assumes there is some file for the package, but only looks for architecture independent packages; is it perhaps finding only sdists and architecture-dependent wheels?

Comment (dnicolodi) [reply .github/workflows/tests-cygwin.yml:39]: Uh? Why do you have wheel files in there? Also, wheel files are only a distribution bundle, they need to be unpacked to be used, so if you install a wheel package you don't have any wheel file at the end, except in a cache, maybe. However, `/usr/share/python-wheels/` is not the default path of the `pip` cache. 

When `venv` creates a new virtual environment it optionally populates it with the `setuptools` and `pip` packages. Because `pip` does not exist yet in the virtual environment, the `ensurepip` module is used for doing it. Customarily, the Python distribution comes with two wheel files for `setuptools` and `pip` bundled with it and these are unpacked by `ensurepip` into the virtual environment. Cygwin decided to patch `ensurepip` to look for the wheel it needs in `/usr/share/python-wheels/` but it does not place any wheel there. How is this supposed to work? Who should place the wheels there?

Now that we understood the issue, I'm patching it in the CI setup. But Python as distribute by Cygwin is broken.

Comment (dnicolodi) [reply .github/workflows/tests-cygwin.yml:39]: The fix https://github.com/mesonbuild/meson-python/blob/f0ce0f55f98128f477246842ca141729f10ecee5/.github/workflows/tests.yml#L117-L128
Now the tests succeed on Cygwin too!

Comment (DWesl) [reply .github/workflows/tests-cygwin.yml:39]: Success!

**Comment by DWesl** [comment]
Superseded by #202

**Pull request closed without merging**
